paddle/fluid/operators/gather_scatter_kernel.cc

```diff
@@ -0,0 +1,203 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/operators/gather_scatter_kernel.h"
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+class TensorAssign {
+ public:
+  template <typename tensor_t>
+  void operator()(tensor_t* self_data, tensor_t* src_data) const {
+    *self_data = *src_data;
+    VLOG(3) << "self_data assigned:" << *self_data;
+  }
+};
+static TensorAssign tensor_assign;
+
+class ReduceAdd {
+ public:
+  template <typename tensor_t>
+  void operator()(tensor_t* self_data, tensor_t* src_data) const {
+    *self_data += *src_data;
+    VLOG(3) << "self_data added:" << *self_data;
+  }
+};
+
+static ReduceAdd reduce_add;
+
+class ReduceMultiply {
+ public:
+  template <typename tensor_t>
+  void operator()(tensor_t* self_data, tensor_t* src_data) const {
+    *self_data *= *src_data;
+  }
+
+  void operator()(bool* self_data, bool* src_data) const {
+    *self_data = *self_data && *src_data;
+  }
+};
+
+static ReduceMultiply reduce_mul;
+
+template <typename tensor_t, typename index_t = int64_t,
+          bool is_scatter_like = true>
+struct cpu_gather_scatter_functor {
+  template <typename func_t>
+  void operator()(Tensor self, int dim, const Tensor& index, const Tensor& src,
```

Comment:
Regular functions that operates on Tensor should take a Tensor either as 
1. const Tensor& (const input);
2. Tensor* (output).

Pass a tensor by-value would cause Tensor assignment, which is a shallow copy. It would have the same effect as `ShareDataWith`, which is not encouraged.

Reply:
I just modified the input tensor in my code, is that allowed?

Reply:
Mutating the input is generally not allowed in opertor kernels.
Operators assume the inputs const tensors, and write the results to the outputs.

Inplace operations are achived by passing the input also as an output parameter.

Reply:
Though pass a tensor by value here does not cause a deep copy, pass a pointer to tensor is much more common in the code base.